Merge py-uproot4 into py-uproot

glennpj opened the pull request:
This PR merges the py-uproot4 package into the py-uproot package.
- Added DEPRECATED comment to py-uproot4
- copied py-uproot4/package.py to py-uproot/package.py
- kept comment from py-uproot
- kept newer version from py-uproot

@vvolkl

glennpj committed:
Merge py-uproot4 into py-uproot

This PR merges the py-uproot4 package into the py-uproot package.
- Added DEPRECATED comment to py-uproot4
- copied py-uproot4/package.py to py-uproot/package.py
- kept comment from py-uproot
- kept newer version from py-uproot

adamjstewart commented on var/spack/repos/builtin/packages/py-uproot/package.py:
```diff
@@ -16,11 +16,36 @@ class PyUproot(PythonPackage):
-    pypi     = "uproot/uproot-4.0.6.tar.gz"
+    homepage = "https://uproot4.readthedocs.io"
+    git      = "https://github.com/scikit-hep/uproot4"
+    url      = "https://github.com/scikit-hep/uproot4/archive/0.0.27.tar.gz"
```
Why switch to GitHub? PyPI is preferred for Python libraries. Also, the checksums may differ.

glennpj replied:
That is what was in the py-uproot4 package.

adamjstewart replied:
Personally I would switch both to PyPI, although I don't care too much about converting deprecated, soon to be removed packages.

glennpj replied:
> Also, the checksums may differ.

Good point. That should only affect the newer version that I kept from the original py-uproot.

glennpj replied:
Looking at the py-awkward an py-awkward1 packages, py-awkward1 uses git and py-awkward uses pypi.

@vvolkl Do we want to standardize on git or pypi for awkward and uproot?

vvolkl replied:
yes, I think we just didn't bother for awkward1 as it was deprecated, but it makes sense to use pypi

adamjstewart replied:
The "official" Spack policy is to standardize on PyPI whenever possible: https://spack.readthedocs.io/en/latest/build_systems/pythonpackage.html#pypi-vs-github

If PyPI only has wheels, then I use GitHub instead.

glennpj replied:
This is back to pypi, and just the latest version. I assume that we keep py-uproot4 using git rather than change that.

glennpj committed:
Fix the checksum for version 4.0.6

This is necessary for the change from pypi to git.

vvolkl commented:
Thanks @glennpj, I think it makes sense to switch to pypi. You could still deprecate the `version`s, like done here: https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/py-awkward1/package.py#L18

glennpj commented:
> Thanks @glennpj, I think it makes sense to switch to pypi. You could still deprecate the `version`s, like done here: https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/py-awkward1/package.py#L18

Okay, should I keep all of the versions from py-uproot4?

vvolkl commented:
> > Thanks @glennpj, I think it makes sense to switch to pypi. You could still deprecate the `version`s, like done here: https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/py-awkward1/package.py#L18
> 
> Okay, should I keep all of the versions from py-uproot4?

just keeping the latest one is fine, in my opinion

glennpj committed:
Add Spack deprecation parameter to each version

glennpj committed:
iterate over versions

- changed from git to pypi
- revert checksum to pypi version
- only keep latest version in py-uproot
- add back python version dependency that got dropped

glennpj committed:
Fix indentation error that crept in

glennpj committed:
Fix copy/paste error

adamjstewart commented on var/spack/repos/builtin/packages/py-uproot/package.py:
```diff
@@ -7,7 +7,7 @@
 
 class PyUproot(PythonPackage):
-    """ROOT I/O in pure Python and NumPy.
+    """ROOT I/O in pure Python and NumPy
```
```suggestion
    """ROOT I/O in pure Python and NumPy.
```
`spack info py-uproot` gets rid of newline chars, so you'll want this period here.

glennpj committed:
Update var/spack/repos/builtin/packages/py-uproot/package.py

Co-authored-by: Adam J. Stewart <ajstewart426@gmail.com>

adamjstewart commented:
Do we actually want to deprecate uproot3/4 and awkward0/1? It seems like some of the packages that @glennpj added actually require these old versions. Or is the plan to keep uproot3 and awkward0 and only deprecate uproot4 and awkward1?

glennpj commented:
@vvolkl Does everything look okay to you?

glennpj commented:
> Do we actually want to deprecate uproot3/4 and awkward0/1? It seems like some of the packages that @glennpj added actually require these old versions. Or is the plan to keep uproot3 and awkward0 and only deprecate uproot4 and awkward1?

My understanding is that py-uproot4 and py-awkward1 are deprecated because they are replaced by py-uproot and py-awkward, respectively. That is not the case for py-uproot3/py-awkward0, which will stand on their own as legacy packages.

vvolkl commented:
> Do we actually want to deprecate uproot3/4 and awkward0/1? It seems like some of the packages that @glennpj added actually require these old versions. Or is the plan to keep uproot3 and awkward0 and only deprecate uproot4 and awkward1?

Yes, I was intending to do the latter, since as you say uproot3/awkward0  may still be needed as legacy packages, and are actually also still maintained as such. But uproot4/awkward1 are just renamed, so we don't want people to use these package names. I'm also fine with just deleting those, since I don't think too many people used them, but we can wait a release now that we deprecated them.

@glennpj LGTM!
